In [1]:
# imports
# packages used
import librosa
import os 

import math
import numpy as np
import pandas as pd
import ast
import re

import dask.array as da
# from sklearn.model_selection import train_test_split

# import torch
# import torch.nn as nn
# import torch.nn.functional as F
# import torch.optim as optim

# from tqdm.notebook import tqdm, trange
# from torch.utils.data import Dataset, DataLoader

# import matplotlib.pyplot as plt

In [2]:
file_path_1 = "mtg-jamendo-dataset/prepare_data/data_01_X.npy"
file_path_2 = "mtg-jamendo-dataset/prepare_data/data_02_X.npy"
file_path_3 = "mtg-jamendo-dataset/prepare_data/data_03_X.npy"
file_path_4 = "mtg-jamendo-dataset/prepare_data/data_04_X.npy"
file_path_5 = "mtg-jamendo-dataset/prepare_data/data_05_X.npy"

files = [file_path_1, file_path_2, file_path_3, file_path_4, file_path_5]
arr_1 = [np.load(file) for file in files]
stacked_arr_1 = np.vstack(arr_1)
print(stacked_arr_1.shape)

# Create a list of lazy Dask arrays from .npy files
# dask_arrays = [da.from_array(np.load(f"mtg-jamendo-dataset/prepare_data/data_0{i}_X.npy", mmap_mode='r'), chunks='auto') for i in range(1, 10, 1)]

# Concatenate the arrays into one large Dask array
# large_dask_array = da.concatenate(dask_arrays, axis=0)

(4866, 204, 1000)


In [10]:
file_1 = "mtg-jamendo-dataset/prepare_data/data_01_y.npy"
file_2 = "mtg-jamendo-dataset/prepare_data/data_02_y.npy"
file_3 = "mtg-jamendo-dataset/prepare_data/data_03_y.npy"
file_4 = "mtg-jamendo-dataset/prepare_data/data_04_y.npy"
file_5 = "mtg-jamendo-dataset/prepare_data/data_05_y.npy"

files_y_1 = [file_1, file_2, file_3, file_4, file_5]
arr_y_1 = [np.load(file, allow_pickle=True) for file in files_y_1]
flat_y_1 = []
for sub in arr_y_1:
    flat_y_1.extend(sub)
print(len(flat_y_1))

4866


In [11]:
# store first 5 groups
np.save("mtg-jamendo-dataset/prepare_data/stack_data_01_X.npy", stacked_arr_1)
np.save("mtg-jamendo-dataset/prepare_data/stack_data_01_y.npy", np.array(flat_y_1))

In [5]:
# part 2
file_path_6 = "mtg-jamendo-dataset/prepare_data/data_06_X.npy"
file_path_7 = "mtg-jamendo-dataset/prepare_data/data_07_X.npy"
file_path_8 = "mtg-jamendo-dataset/prepare_data/data_08_X.npy"
file_path_9 = "mtg-jamendo-dataset/prepare_data/data_09_X.npy"
file_path_10 = "mtg-jamendo-dataset/prepare_data/data_10_X.npy"

files = [file_path_6, file_path_7, file_path_8, file_path_9, file_path_10]
arr_2 = [np.load(file) for file in files]
stacked_arr_2 = np.vstack(arr_2)
print(stacked_arr_2.shape)

(4597, 204, 1000)


In [6]:
file_6 = "mtg-jamendo-dataset/prepare_data/data_06_y.npy"
file_7 = "mtg-jamendo-dataset/prepare_data/data_07_y.npy"
file_8 = "mtg-jamendo-dataset/prepare_data/data_08_y.npy"
file_9 = "mtg-jamendo-dataset/prepare_data/data_09_y.npy"
file_10 = "mtg-jamendo-dataset/prepare_data/data_10_y.npy"

files_y_2 = [file_6, file_7, file_8, file_9, file_10]
arr_y_2 = [np.load(file, allow_pickle=True) for file in files_y_2]
flat_y_2 = []
for sub in arr_y_2:
    flat_y_2.extend(sub)
print(len(flat_y_2))

4597


In [7]:
# store first 5 groups
np.save("mtg-jamendo-dataset/prepare_data/stack_data_02_X.npy", stacked_arr_2)
np.save("mtg-jamendo-dataset/prepare_data/stack_data_02_y.npy", np.array(flat_y_2))

In [8]:
# part 3
file_path_11 = "mtg-jamendo-dataset/prepare_data/data_11_X.npy"
file_path_12 = "mtg-jamendo-dataset/prepare_data/data_12_X.npy"
file_path_13 = "mtg-jamendo-dataset/prepare_data/data_13_X.npy"
file_path_14 = "mtg-jamendo-dataset/prepare_data/data_14_X.npy"
file_path_15 = "mtg-jamendo-dataset/prepare_data/data_15_X.npy"

files = [file_path_11, file_path_12, file_path_13, file_path_14, file_path_15]
arr_3 = [np.load(file) for file in files]
stacked_arr_3 = np.vstack(arr_3)
print(stacked_arr_3.shape)

(4557, 204, 1000)


In [9]:
file_11 = "mtg-jamendo-dataset/prepare_data/data_11_y.npy"
file_12 = "mtg-jamendo-dataset/prepare_data/data_12_y.npy"
file_13 = "mtg-jamendo-dataset/prepare_data/data_13_y.npy"
file_14 = "mtg-jamendo-dataset/prepare_data/data_14_y.npy"
file_15 = "mtg-jamendo-dataset/prepare_data/data_15_y.npy"

files_y_3 = [file_11, file_12, file_13, file_14, file_15]
arr_y_3 = [np.load(file, allow_pickle=True) for file in files_y_3]
flat_y_3 = []
for sub in arr_y_3:
    flat_y_3.extend(sub)
print(len(flat_y_3))

4557


In [10]:
# store first 5 groups
np.save("mtg-jamendo-dataset/prepare_data/stack_data_03_X.npy", stacked_arr_3)
np.save("mtg-jamendo-dataset/prepare_data/stack_data_03_y.npy", np.array(flat_y_3))

In [2]:
# combine three 
d1 = np.load("mtg-jamendo-dataset/prepare_data/stack_data_01_X.npy")
d2 = np.load("mtg-jamendo-dataset/prepare_data/stack_data_02_X.npy")
d3 = np.load("mtg-jamendo-dataset/prepare_data/stack_data_03_X.npy")
stack = np.vstack([d1, d2, d3])
print(stack.shape)

(14020, 204, 1000)


In [3]:
d1_y = np.load("mtg-jamendo-dataset/prepare_data/stack_data_01_y.npy")
d2_y = np.load("mtg-jamendo-dataset/prepare_data/stack_data_02_y.npy")
d3_y = np.load("mtg-jamendo-dataset/prepare_data/stack_data_03_y.npy")
stack_y = []
for sub in [d1_y, d2_y, d3_y]:
    stack_y.extend(sub)
print(len(stack_y))

14020


In [4]:
np.save("mtg-jamendo-dataset/prepare_data/large_X.npy", stack)
np.save("mtg-jamendo-dataset/prepare_data/large_y.npy", np.array(stack_y))

In [2]:
X = np.load("mtg-jamendo-dataset/prepare_data/large_X.npy")
y = np.load("mtg-jamendo-dataset/prepare_data/large_y.npy")

In [3]:
print(X.shape)
print(y.shape)

(14020, 204, 1000)
(14020,)


## Original 8071 timesteps

In [17]:
large_dask_array

dask.array<concatenate, shape=(8547, 204, 8071), dtype=float64, chunksize=(286, 204, 286), chunktype=numpy.ndarray>

In [18]:
memory_usage_bytes = large_dask_array.size * large_dask_array.dtype.itemsize
print("Estimated memory usage in bytes:", memory_usage_bytes)

Estimated memory usage in bytes: 112579989984


In [22]:
label_counts = {}

# Assuming file naming as 'data_xx_y.npy' for labels
for i in range(1, 16):  # 15 files as per your description
    labels = np.load(f'mtg-jamendo-dataset/prepare_data/data_{i:02d}_y.npy', allow_pickle=True)
    unique, counts = np.unique(labels, return_counts=True)
    for label, count in zip(unique, counts):
        if label in label_counts:
            label_counts[label] += count
        else:
            label_counts[label] = count

# Find the minimum count to balance the dataset
min_count = min(label_counts.values())
print("Minimum samples per class needed:", min_count)

Minimum samples per class needed: 13


In [31]:
label_counts.values()

dict_values([328, 288, 503, 348, 368, 239, 309, 283, 418, 187, 179, 218, 878, 462, 220, 205, 150, 606, 608, 839, 375, 691, 247, 69, 115, 96, 557, 78, 13, 57, 74, 143, 431, 364, 152, 44, 583, 48, 96, 65, 107, 69, 38, 440, 70, 127, 243, 40, 100, 44, 228, 201, 62, 61, 29, 57, 73, 65, 32])

In [32]:
label_counts

{0: 328,
 1: 288,
 2: 503,
 3: 348,
 4: 368,
 5: 239,
 6: 309,
 7: 283,
 8: 418,
 9: 187,
 10: 179,
 11: 218,
 12: 878,
 13: 462,
 14: 220,
 15: 205,
 16: 150,
 17: 606,
 18: 608,
 19: 839,
 20: 375,
 22: 691,
 23: 247,
 24: 69,
 25: 115,
 26: 96,
 27: 557,
 28: 78,
 29: 13,
 30: 57,
 31: 74,
 32: 143,
 33: 431,
 34: 364,
 35: 152,
 36: 44,
 37: 583,
 38: 48,
 39: 96,
 40: 65,
 41: 107,
 42: 69,
 43: 38,
 44: 440,
 45: 70,
 46: 127,
 47: 243,
 48: 40,
 49: 100,
 50: 44,
 51: 228,
 52: 201,
 53: 62,
 54: 61,
 55: 29,
 57: 57,
 58: 73,
 21: 65,
 56: 32}

## Issues

1. We have 15 .npy files for X and 15 corresponding .npy files for y. However, we can't apply np.load to successfully load these files due to their large RAM usage.

    solution: right now, each .npy file is roughly in the size (900, 204, 8071). Try with the first .npy generating process, to make the .npy file in the size (900, 204, 1000) to see what might happen.